features/environment.py

```diff
@@ -98,36 +97,70 @@ def from_environ(cls) -> "UAClientBehaveConfig":
 
 
 def before_all(context: Context) -> None:
-    """behave will invoke this before anything else happens.
-
-    In this function, we launch a container, install ubuntu-advantage-tools and
-    then capture an image.  This image is then reused by each feature, reducing
-    test execution time.
-    """
+    """behave will invoke this before anything else happens."""
     userdata = context.config.userdata
     context.reuse_container = userdata.get("reuse_container")
+    context.series_image_name = {}
+    context.series_reuse_image = ""
     context.config = UAClientBehaveConfig.from_environ()
-    if context.config.reuse_image is None:
-        create_trusty_uat_lxd_image(context)
-    else:
-        context.image_name = context.config.reuse_image
+    if context.config.reuse_image:
+        series = lxc_get_series(context.config.reuse_image, image=True)
+        if series is not None:
+            context.series_reuse_image = series
+        else:
+            print(" Could not check image series. It will not be used. ")
+            context.config.reuse_image = None
 
 
-def before_scenario(context: Context, scenario: Scenario):
-    for tag in scenario.effective_tags:
+def before_feature(context: Context, feature: Feature):
+    for tag in feature.tags:
         parts = tag.split(".")
         if parts[0] == "uses":
             val = context
             for attr in parts[1:]:
                 val = getattr(val, attr, None)
                 if val is None:
-                    scenario.skip(
+                    feature.skip(
                         reason="Skipped because tag value was None: {}".format(
                             tag
                         )
                     )
 
 
+def before_scenario(context: Context, scenario: Scenario):
+    """
+    In this function, we launch a container, install ubuntu-advantage-tools and
+    then capture an image. This image is then reused by each scenario, reducing
+    test execution time.
+    """
+    for tag in scenario.effective_tags:
+        parts = tag.split(".")
+        if parts[0] == "series":
+            series = parts[1]
+            if context.config.reuse_image is None:
+                if series not in context.series_image_name:
+                    create_uat_lxd_image(context, series)
+            else:
+                if series == context.series_reuse_image:
+                    context.series_image_name[
+                        series
+                    ] = context.config.reuse_image
```

Comment:
I don't think we should be doing this on every scenario, let's set this once up in before_all if we can.

Reply:
`before_all` hook doesn't have `feature` or `scenario` parameter, so it doesn't have any info about the tags. and `series` come from `scenario.effective_tags`.